overlord/snapstate/target.go

```diff
@@ -0,0 +1,588 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016-2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snapstate
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/snapcore/snapd/asserts/snapasserts"
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/snap/naming"
+	"github.com/snapcore/snapd/store"
+)
+
+// Options contains optional parameters for the snapstate operations. All of
+// these fields are optional and can be left unset. The options in this struct
+// apply to all snaps that are part of an operation. Options that apply to
+// individual snaps can be found in RevisionOptions.
+type Options struct {
+	// Flags contains flags that apply to the entire operation.
+	Flags Flags
+	// UserID is the ID of the user that is performing the operation.
+	UserID int
+	// DeviceCtx is an optional device context that will be used during the
+	// operation.
+	DeviceCtx DeviceContext
+	// PrereqTracker is an optional prereq tracker that will be used to keep
+	// track of all snaps (explicitly requested and implicitly required snaps)
+	// that might need to be installed during the operation.
+	PrereqTracker PrereqTracker
+	// FromChange is the change that triggered the operation.
+	FromChange string
+	// Seed should be true while seeding the device. This indicates that we
+	// shouldn't require that the device is seeded before installing/updating
+	// snaps.
+	Seed bool
+	// RequireOneSnap is a boolean flag indicating that this operation is expected
+	// to only operate on one snap (excluding any prerequisite snaps that may be
+	// required). If this is true, then the operation will fail if more than one
+	// snap is being operated on. This flag primarily exists to support the
+	// pre-existing behavior of calling InstallMany with one snap vs calling
+	// Install.
+	RequireOneSnap bool
```

Comment:
wouldn't ExpectOneSnap be clearer for this? OTOH I would also be ok with calling this SingleSnap